packages/signers/src/sign-transaction.ts

```diff
@@ -0,0 +1,202 @@
+    );
+
+    // Identify the modifying signers from the other signers.
+    const modifyingSigners = identifyTransactionModifyingSigners(otherSigners);
+
+    // Use any remaining signers as partial signers.
+    const partialSigners = otherSigners
+        .filter(isTransactionPartialSigner)
+        .filter(signer => !(modifyingSigners as typeof otherSigners).includes(signer));
+
+    return Object.freeze({ modifyingSigners, partialSigners, sendingSigner });
+}
+
+/** Identifies the best signer to use as a TransactionSendingSigner, if any */
+function identifyTransactionSendingSigner(signers: readonly TransactionSigner[]): TransactionSendingSigner | null {
+    // Ensure there are any TransactionSendingSigners in the first place.
+    const sendingSigners = signers.filter(isTransactionSendingSigner);
+    if (sendingSigners.length === 0) return null;
+
+    // Prefer sending signers that do not offer other interfaces.
+    const sendingOnlySigners = sendingSigners.filter(
+        signer => !isTransactionModifyingSigner(signer) && !isTransactionPartialSigner(signer)
+    );
+    if (sendingOnlySigners.length > 0) {
+        return sendingOnlySigners[0];
+    }
+
+    // Otherwise, choose any sending signer.
+    return sendingSigners[0];
+}
+
+/** Identifies the best signers to use as TransactionModifyingSigners, if any */
+function identifyTransactionModifyingSigners(
+    signers: readonly (TransactionModifyingSigner | TransactionPartialSigner)[]
+): readonly TransactionModifyingSigner[] {
+    // Ensure there are any TransactionModifyingSigner in the first place.
+    const modifyingSigners = signers.filter(isTransactionModifyingSigner);
+    if (modifyingSigners.length === 0) return [];
+
+    // Prefer modifying signers that do not offer partial signing.
+    const nonPartialSigners = modifyingSigners.filter(signer => !isTransactionPartialSigner(signer));
+    if (nonPartialSigners.length > 0) return nonPartialSigners;
+
+    // Otherwise, choose only one modifying signer (whichever).
+    return [modifyingSigners[0]];
+}
+
+/**
+ * Signs a transaction using the provided TransactionModifyingSigners
+ * sequentially followed by the TransactionPartialSigners in parallel.
+ */
+async function signModifyingAndPartialTransactionSigners<
+    TTransaction extends CompilableTransactionWithSigners = CompilableTransactionWithSigners
+>(
+    transaction: TTransaction,
+    modifyingSigners: readonly TransactionModifyingSigner[] = [],
+    partialSigners: readonly TransactionPartialSigner[] = []
+): Promise<TTransaction & ITransactionWithSignatures> {
+    // Handle modifying signers sequentially.
+    const modifiedTransaction = await modifyingSigners.reduce(async (transaction, modifyingSigner) => {
+        const [tx] = await modifyingSigner.modifyAndSignTransactions([await transaction]);
+        return Object.freeze(tx);
+    }, Promise.resolve(transaction) as Promise<TTransaction>);
```

Comment:
I may be missing something, but how does this not void any signatures before the final `modifyAndSign`?

Reply:
It may invalidate previous signatures but it's up to the signer itself to decide if they should modify when they encounter a message or a transaction that already contains signatures.

Here's a more detailed description of this from the internal document:

---

### Handling modifying conflicts

The first type of conflict occurs when multiple message or transaction modifying signers are provided.

For instance, say we have two `TransactionModifyingSigner` A and B for the same transaction. If A signs before B, B will potentially modify the transaction again which invalidates A’s signature. The same scenario occurs if B signs before A.

There are however two ways to get around these conflicts and, as such, we shouldn’t fail if we encounter this scenario.

- **The Signer may decide to not modify**. In our previous example, A or B could first check the state of the transaction before deciding if they want to modify the transaction or not. Typically, if other signatures exist on the given transaction, they will fallback to signing without modifying it.
- **The Signer may allow the app to decide**. A signer may implement both the `TransactionPartialSigner` and `TransactionModifyingSigner` interfaces (or equivalent for messages) and let the application decide which one they want to use for their specific use-cases. The tradeoff for the app would be that `TransactionPartialSigner` brings more flexibility and can be executed in parallel whereas `TransactionModifyingSigner` may bring additional features for the app.

Now, if none of these workarounds are applicable and both A and B *must* modify the transaction before signing it, then we should still not fail and let the second wallet itself throw an error explaining why it cannot successfully modify and sign the given transaction.

Reply:
So you're basically saying the onus is on the second wallet for making sure they aren't modifying an already signed transaction? And if they are, but this transaction comes signed from some SDK's method, that's the fault of the first wallet, yeah? I just want to make sure we're not enabling this confusion/collison.

Reply:
> So you're basically saying the onus is on the second wallet for making sure they aren't modifying an already signed transaction?

Yes, it's the responsibility of the subsequent wallets to decide whether they can also modify the transaction or not. If they _must_ modify the transaction and cannot then they can throw an error explaining why they cannot sign without invalidating previous signatures. But if _we_ throw an error on our helper methods, then we stop supporting valid use-cases. And our goal with this Signer API is that **it should not prevent wallets to do what they could do without**.

> that's the fault of the first wallet, yeah?

I don't think it's a wallet's fault in particular but just a combination of Signers that end up in an impossible scenario (with or without the Signer API). But it's the responsibility of the wallet that identifies the issue to fail accordingly.